drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1181 @@
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+/*
+ * Destroys the front and back buffers of a certain output and then destroys
+ * the output itself.
+ */
+
+static void modeset_output_destroy(int fd, struct modeset_output *out)
+{
+	/* destroy front/back framebuffers */
+	modeset_destroy_fb(fd, &out->bufs[0]);
+	modeset_destroy_fb(fd, &out->bufs[1]);
+
+	free(out);
+}
+
+/*
+ * With a certain combination of connector+CRTC, we look for a suitable primary
+ * plane for it. After that, we retrieve connector, CRTC and plane objects
+ * properties from the device. These objects are used during the atomic modeset
+ * setup (see modeset_atomic_prepare_commit()) and also during the page-flips
+ * (see modeset_draw_out() and modeset_atomic_commit()).
+ *
+ * Besides that, we have to create a blob property that receives the output
+ * mode. When we perform an atomic commit, the driver expects a CRTC property
+ * named "MODE_ID", which points to the id of a blob. This usually happens for
+ * properties that are not simple types. In this particular case, out->mode is a
+ * struct. But we could have another property that expects the id of a blob that
+ * holds an array, for instance.
+ */
+
+static struct modeset_output *modeset_output_create(int fd, drmModeRes *res,
+						    drmModeConnector *conn)
+{
+	int ret;
+	struct modeset_output *out;
+
+	/* creates an output structure */
+	out = malloc(sizeof(*out));
+	memset(out, 0, sizeof(*out));
+	out->conn_id = conn->connector_id;
+
+	/* check if a monitor is connected */
+	if (conn->connection != DRM_MODE_CONNECTED) {
+		fprintf(stderr, "ignoring unused connector %u\n",
+			conn->connector_id);
+		goto out_error;
+	}
+
+	/* check if there is at least one valid mode */
+	if (conn->count_modes == 0) {
+		fprintf(stderr, "no valid mode for connector %u\n",
+			conn->connector_id);
+		goto out_error;
+	}
+
+	/* copy the mode information into our output structure */
+	memcpy(&out->mode, &conn->modes[0], sizeof(out->mode));
+	/* create the blob property using out->mode and save its id in the output*/
+	if (drmModeCreatePropertyBlob(fd, &out->mode, sizeof(out->mode),
+	                              &out->mode_blob_id) != 0) {
+		fprintf(stderr, "couldn't create a blob property\n");
+		goto out_error;
+	}
+	fprintf(stderr, "mode for connector %u is %ux%u\n",
+	        conn->connector_id, out->bufs[0].width, out->bufs[0].height);
+
+	/* find a crtc for this connector */
+	ret = modeset_find_crtc(fd, res, conn, out);
+	if (ret) {
+		fprintf(stderr, "no valid crtc for connector %u\n",
+			conn->connector_id);
+		goto out_error;
+	}
+
+	/* with a connector and crtc, find a primary plane */
+	ret = modeset_find_plane(fd, out);
+	if (ret) {
+		fprintf(stderr, "no valid plane for crtc %u\n", out->crtc_id);
+		goto out_error;
+	}
+
+	/* gather properties of our connector, CRTC and planes */
+	ret = modeset_setup_objects(fd, out);
+	if (ret) {
+		fprintf(stderr, "cannot get plane properties\n");
+		goto out_error;
+	}
+
```

Comment:
From this point onwards, a failure should also release what `modeset_setup_objects()` allocated.

Reply:
FIxed.